File: Objective-C-Project/Practicing/TestStoryBoard/TestStoryBoard/PlayTraditionalViewController.m
//
//  PlayTriditionalViewController.m
//  TestStoryBoard
//
//  Created by 谢乾坤 on 2/18/16.
//  Copyright © 2016 QiankunXie. All rights reserved.
//

#import "PlayTraditionalViewController.h"
#import "QStack.h"

@interface PlayTraditionalViewController ()

@property (weak, nonatomic) IBOutlet UILabel *questionLabel;

@property (weak, nonatomic) IBOutlet UIView *questionLabelView;

@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray *answers;

@property (strong, nonatomic) IBOutletCollection(UIView) NSArray *answersView;

@property (weak, nonatomic) IBOutlet UIView *timeBarView;

@property (weak, nonatomic) IBOutlet UIView *timeBar;

@property (weak, nonatomic) IBOutlet UILabel *timeBarClock;

//for timer
@property (nonatomic, strong) NSTimer *timeBarTimer;

@property (nonatomic) NSNumber *timeNow;

@property (nonatomic) NSInteger time;


@property (nonatomic) NSInteger currentQuestionNumber;
@property (nonatomic) BOOL isAllowSubmit;
@property (nonatomic) NSInteger currentSumbittedAnswer;


@end

@implementation PlayTraditionalViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    [self setUIBeforeStart];
}

- (void) viewWillAppear:(BOOL)animated{
    
    [self startGame];
}



- (void)viewWillDisappear:(BOOL)animated
{
    [[QStack sharedQStack] endSocket];
    
    
}

- (void)setUIBeforeStart{
    
    self.view.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"Background.png"]];

    
    for(UIView *view in self.answersView){
        
        [view.layer setBorderColor:[UIColor whiteColor].CGColor];
        view.layer.cornerRadius = 5;
        view.layer.masksToBounds = YES;
        [view.layer setBorderWidth:2.0f];
        [view setBackgroundColor:[UIColor clearColor]];
    }
    
    [self.questionLabelView.layer setBorderColor:[UIColor whiteColor].CGColor];
    self.questionLabelView.layer.cornerRadius = 5;
    self.questionLabelView.layer.masksToBounds = YES;
    [self.questionLabelView.layer setBorderWidth:2.0f];
    [self.questionLabelView setBackgroundColor:[UIColor clearColor]];
    
    //set time bar
    [self.timeBarView.layer setBorderColor:[UIColor whiteColor].CGColor];
    self.timeBarView.layer.cornerRadius = 17;
    self.timeBarView.layer.masksToBounds = YES;
    [self.timeBarView.layer setBorderWidth:2.0f];
    [self.timeBarView setBackgroundColor:[UIColor clearColor]];
    
    self.timeBar.layer.cornerRadius = 17;
    self.timeBar.layer.masksToBounds = YES;
    
    self.timeBarClock.layer.cornerRadius = 27;
    self.timeBarClock.layer.masksToBounds = YES;
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}
- (void)startGame {
    
    
    QStack *qstack = [QStack sharedQStack];
    
    [qstack connectGameServer:self.uuid completionHandler:^(NSError *error) {
        if (!error){
            NSLog(@"start loading data");
            [qstack startSocket:^(NSDictionary *data) {
                
                [self choosePayLoadType:data];
                
            }];
        }
    }];

    
}
- (void) choosePayLoadType:(NSDictionary *)data
{
    NSString* type = data[@"type"];
    
    NSDictionary * gameInfo = data[@"payload" ];
    
    if ([type isEqual: @"sendQuestion"]){
        
        self.questionLabel.lineBreakMode = NSLineBreakByWordWrapping;
        self.questionLabel.numberOfLines = 0;
        self.currentSumbittedAnswer = -1;

        
        //set current question number && submit allowance
        self.currentQuestionNumber = [gameInfo[@"questionNum"] intValue];
        self.isAllowSubmit = YES;
        
        NSString * s = [self fixString:gameInfo[@"question"]];
        
        [self.questionLabel setText:[NSString stringWithFormat:@"%@",s]];
        
        //set timer
        self.time = 5;
        self.timeNow = [NSNumber numberWithDouble:self.time];
        [self.timeBarClock setText:[NSString stringWithFormat:@"%ld",(long)self.time]];

        
        //game answers
        NSArray *receivedAnswers = gameInfo[@"answers"];
        
        for(UIButton *btn in self.answers){
            //reset button color
            NSInteger index = [self.answers indexOfObject:btn];
            NSString *s = receivedAnswers[index];
            [btn setTitle:s forState:UIControlStateNormal];
            [btn setBackgroundColor:[UIColor clearColor]];
        }
        //control time delay for answer display
        double delayInSeconds = 0.0;
        dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds * NSEC_PER_SEC));
        dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
            //start timer
            self.timeBarTimer = [NSTimer scheduledTimerWithTimeInterval:0.05
                                                                 target:self
                                                               selector:@selector(startTimer:)
                                                               userInfo:nil
                                                                repeats:YES];
            
        });

        
    }
    if ([type isEqual: @"correctAnswer"]){
        
        NSInteger answerNum = [gameInfo[@"correctAnswer"] integerValue];
        self.isAllowSubmit = NO;
        
        //stop timer
        [self.timeBarTimer invalidate];
        self.timeBarTimer = nil;
        
        if ([gameInfo[@"questionNum"] intValue] == self.currentQuestionNumber) {

            if (self.currentSumbittedAnswer >= 0){
                if(self.currentSumbittedAnswer == answerNum)
                    [self.answers[answerNum] setBackgroundColor:[UIColor greenColor]];
                else {
                    [self.answers[answerNum] setBackgroundColor:[UIColor greenColor]];
                    [self.answers[self.currentSumbittedAnswer] setBackgroundColor:[UIColor redColor]];
                }
            } else {
                [self.answers[answerNum] setBackgroundColor:[UIColor greenColor]];
            }
        
        }
        
    }
    
    if([type isEqual:@"gameResult"]){
        
        [self.questionLabel setText:@"Game Over"];
        
        for(UIButton *btn in self.answers){
            [btn setTitle:@"" forState:UIControlStateNormal];
            [btn setBackgroundColor:[UIColor clearColor]];
        }
        [self.timeBar setBackgroundColor:[UIColor clearColor]];
    }
    
}

// for time bar
- (void)startTimer:(NSTimer *)timer{
    
    double tempTimeNow = [self.timeNow doubleValue] - 0.05;
    
    self.timeNow = [NSNumber numberWithDouble:tempTimeNow];
    
    [self.timeBarClock setText:[NSString stringWithFormat:@"%i",(int)tempTimeNow]];
    
    CGRect newRect = self.timeBar.frame;
    
    double ratio = 1.0 * (self.time - tempTimeNow)/self.time;
    
    newRect.origin.x = ratio * (self.timeBarView.frame.size.width - self.timeBarClock.frame.size.width) / 2;
    
    newRect.size.width = self.timeBarView.frame.size.width - 2* newRect.origin.x;
    
    [self.timeBar setFrame:newRect];
    
}


- (NSString *)fixString:(NSString *)neededFixString
{
    NSString* s = [neededFixString stringByReplacingOccurrencesOfString:@"<q>" withString:@"\""];
    s = [s stringByReplacingOccurrencesOfString:@"<a>" withString:@"\'"];
    
    return s;
}

- (IBAction)chooseAnswer:(id)sender {
    
    if (self.isAllowSubmit) {
        
        NSInteger index = [self.answers indexOfObject:sender];
        //record current submitted answer
        self.currentSumbittedAnswer = index;
        
        self.isAllowSubmit = NO;
        
        //stop timer
        [self.timeBarTimer invalidate];
        self.timeBarTimer = nil;
        
        NSLocale* currentLocale = [NSLocale currentLocale];
        
        NSDictionary* answer = @{
                                 @"questionNum" : [NSNumber numberWithInteger:self.currentQuestionNumber],
                                 @"answer" : [NSNumber numberWithInteger:index],
                                 @"time" : [[NSDate date] descriptionWithLocale:currentLocale]
                                 
                                 };
        [[QStack sharedQStack] submitAnswer:answer];
        
    }
    

    
    
}



/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: Swift-Project/GStack-Swift-SDK/GStack-Swift-SDK/GStack-Swift-SDK/PlayTraditionalViewController.swift
//
//  PlayTraditionalViewController.swift
//  GStack-Swift-SDK
//
//  Created by 谢乾坤 on 3/14/16.
//  Copyright © 2016 QiankunXie. All rights reserved.
//

import UIKit

class PlayTraditionalViewController: UIViewController,GStackGameDelegate {

    
    var tournament: GStackTournament?
    var game: GStackGame?
    
    //for ui
    //for question label
    @IBOutlet weak var questionLabel: UILabel!
    @IBOutlet weak var questionLabelView: UIView!
    
    //for answers
    @IBOutlet var answers: [UIButton]!
    @IBOutlet var answersView: [UIView]!
    
    //for timebar
    @IBOutlet weak var timeBarBackGround: UIView!
    @IBOutlet weak var timeBar: UIView!
    
    @IBOutlet weak var clock: UIView!
    @IBOutlet weak var clockLabel: UILabel!
    
    //for game control
    
    var isAllowSubmit = false
    
    var currentSumbittedAnswer = -1
    
    var currentQuestion: GStackGameQuestion?
    
    var playersNames: [String]?
    
    var isNeedToForfeit = true
    
    //timer
    var timeBarTimer: NSTimer?
    var timeNow: Double?
    var time: Int?
    
    
    /**
     Prepare to start the game and set the GStackGame Instance's delegate to self
     */
    override func viewDidLoad() {
        super.viewDidLoad()
        self.UIInit()
        
        GStack.sharedInstance.GStackStartGameForTournament(tournament!) { (error, game) -> Void in
            
            //prepare for starting game
            self.game = game
            game?.delegate = self
            game?.startGame()
            
        }
        
    }
    
    override func viewWillDisappear(animated: Bool) {
        //use endgame func to disconnect with server
        game?.endGame()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //MARK: - setup UI
    func UIInit() {
        
        //set background
        self.view.backgroundColor = UIColor(patternImage: UIImage(named: "Background.png")!)

        //set question label
        self.questionLabelView.layer.borderColor = UIColor.whiteColor().CGColor
        self.questionLabelView.layer.cornerRadius = 8
        self.questionLabelView.layer.masksToBounds = true
        self.questionLabelView.layer.borderWidth = 2.0
        self.questionLabelView.backgroundColor = UIColor.clearColor()
        self.questionLabel.lineBreakMode = NSLineBreakMode.ByWordWrapping
        //set answers
        
        for view in self.answersView {
            
            view.layer.borderColor = UIColor.whiteColor().CGColor
            view.layer.cornerRadius = 8
            view.layer.masksToBounds = true
            view.layer.borderWidth = 2.0
            view.backgroundColor = UIColor.clearColor()
            self.questionLabel.numberOfLines = 0
        }
        
        //set time bar and clock
        self.timeBarBackGround.layer.borderColor = UIColor.whiteColor().CGColor
        self.timeBarBackGround.layer.cornerRadius = 18
        self.timeBarBackGround.layer.masksToBounds = true
        self.timeBarBackGround.layer.borderWidth = 2.0
        self.timeBarBackGround.backgroundColor = UIColor.clearColor()
        
        self.timeBar.layer.cornerRadius = 18
        self.timeBar.layer.masksToBounds = true
        
        self.clock.layer.cornerRadius = 22
        self.clock.layer.masksToBounds = true
        
    }
    
    
    func setupQuestionUI() {
        
        //clear button text
        for btn in self.answers {
            btn.setTitle("", forState: .Normal)
        }
        for view in self.answersView {
            view.backgroundColor = UIColor.clearColor()
        }
        
        //reset the timebar
        var newRect = self.timeBarBackGround.frame
        newRect.origin = CGPointZero
        self.timeBar.frame = newRect
    }
    

    
    //MARK: - helper func and action
    //for time bar
    
    func startTimer() {
        
        //print("call me")
        
        let tempTimeNow = self.timeNow! - 0.05
        self.timeNow = tempTimeNow
        
        self.clockLabel.text = String(Int(tempTimeNow))
        
        var newRect = self.timeBar.frame
        let ratio = 1.0 * (Double(self.time!) - tempTimeNow)/Double(self.time!)
        
        newRect.origin.x = CGFloat(ratio * Double(self.timeBarBackGround.frame.size.width - self.clock.frame.size.width) / 2)
        
        newRect.size.width = CGFloat(self.timeBarBackGround.frame.size.width - 2 * newRect.origin.x)
        
        self.timeBar.frame = newRect
        
        
    }

    
    @IBAction func backToMain(sender: AnyObject) {
        
        //we need to dialog box asking if the user want to forfeit, and if so we send a forefeit message to the server
        if isNeedToForfeit {
        
        let title = "Back to Main"
        let message = "Are you sure"
        
        let ac = UIAlertController(title: title, message: message, preferredStyle: .ActionSheet)
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .Cancel, handler: nil)
        ac.addAction(cancelAction)
        
        let deleteAction = UIAlertAction(title: "Forfeit", style: UIAlertActionStyle.Destructive, handler: {
            (action) -> Void in
            
            //send forfeit message
            self.game?.sendForfeitMessage()
            
            self.dismissViewControllerAnimated(true, completion: { () -> Void in
                self.game = nil
                self.tournament = nil
            })
            // self.performSegueWithIdentifier("Main", sender: self)
            
        })
        ac.addAction(deleteAction)
        
        //present the alert controller
        presentViewController(ac, animated: true, completion: nil)
            
        } else {
            self.dismissViewControllerAnimated(true, completion: nil)
        }
        
    }
 
    
    @IBAction func submitAnswer(sender: UIButton) {
        
        if self.isAllowSubmit {
            
            let index:Int = self.answers.indexOf(sender)!
            
            self.currentSumbittedAnswer = index
            
            self.game?.submitAnswerForQuestion(self.currentQuestion!, answerIndex: index)
            
            self.isAllowSubmit = false
        }
    }
    
    
    
    // MARK: - Delegation part

    
    func gameDidStart(){
        print("Game did start")
    }
    func willAttemptToReconnect(options: PrimusReconnectOptions) {
        print("Will attempt to reconnect")
    }
    func didReconnect() {
        print("Did reconnect")
    }
    func didLoseConnection() {
        print("Did lose connection")
    }
    func didEstablishConnection() {
        print("Did establish connection")
    }
    func didEncounterError(error: NSError) {
        print("Did encounter error: \(error)")
    }
    func connectionDidEnd() {
        print("Connection did end")
    }
    func connectionDidClose() {
        print("Connection did close")
    }
    func didReceiveTimer(timer:GStackGameTimer){
        //for helping the ui
    }
    func didReceiveErrorMsg(error:GStackGameErrorMsg){
        print("Receive an error: \(error.error)")
    }
    func didReceiveStartRound( round: GStackGameStartRound) {
        
        //clear all the ui
        self.setupQuestionUI()
        self.clockLabel.text = ""
        
    }
    func didReceiveRoundResult( win: GStackGameRoundResult) {
        
    }
    func didReceiveQuestion(question: GStackGameQuestion){
        
        self.setupQuestionUI()
        
        //reset the submitted answer
        self.currentSumbittedAnswer = -1
        self.isAllowSubmit = false
        self.currentQuestion = question
        
        //set question
        self.questionLabel.text = question.formattedQuestion()!
        
        //set timer
        self.time = (question.timer as? Int)! / 1000
        self.timeNow = Double(self.time!)
        
        self.clockLabel.text = String(self.time!)
        
        let delayInSeconds = Double(question.promptTime!)
        let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(delayInSeconds / 1000 * Double(NSEC_PER_SEC)))
        dispatch_after(delayTime, dispatch_get_main_queue()) {
            
            let ans = question.formattedAnswers()
            //set answers' labels
            for i in 0 ..< 4 {
                self.answers[i].setTitle(ans[i], forState: .Normal)
            }
            self.isAllowSubmit = true
            
            //set timer for timebar
            self.timeBarTimer = NSTimer.scheduledTimerWithTimeInterval(0.05, target: self, selector: "startTimer", userInfo: nil, repeats: true)
            
        }
        
        
    }
    func didReceivePlayerInfo(playerInfo: GStackGamePlayerInfo){
        
        var info = [String]()
        for array in playerInfo.teamNames! {
            info.append(array[0])
        }
        self.playersNames = info
        
    }
    func didReceiveCorrectAnswer(correctAnswer: GStackGameCorrectAnswer){
        
        self.timeBarTimer?.invalidate()
        
        self.isAllowSubmit = false
        
        let correct = correctAnswer.correctAnswer as! Int
        
        self.answersView[correct].backgroundColor = UIColor.greenColor()
        
        if self.currentSumbittedAnswer >= 0 {
            
            if self.currentSumbittedAnswer != correct {
                self.answersView[self.currentSumbittedAnswer].backgroundColor = UIColor.redColor()
            }
        }
    }
    func didReceiveUpdatedScore(updatedScore: GStackGameUpdatedScore){
        //implement latter
    }
    func didReceiveGameResult(result: GStackGameResult){

        self.setupQuestionUI()
        self.questionLabel.text = "Game Over"
        
        self.isNeedToForfeit = false
        
        
        //auto go back to main
        let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(3 * Double(NSEC_PER_SEC)))
        dispatch_after(delayTime, dispatch_get_main_queue()) {
            
            self.dismissViewControllerAnimated(true, completion: { () -> Void in
                self.game = nil
                self.tournament = nil
            })
        }
        
    }
    func didReceiveOtherGameFinished(alive:GStackGameOtherGameFinished){
        
        
    }
}
